### PR TITLE
[Fix #5436] Store extending result in all backends

### LINKED ISSUE #5436: Celery 4.3 result_extended doesn't push extended meta data to some backends when set to true
# Checklist

- [x] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.
- [x] I have included the contents of ``pip freeze`` in the issue.
- [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**:

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
celery -A celeryacl report

software -> celery:4.3.0 (rhubarb) kombu:4.5.0 py:3.6.8
billiard:3.6.0.0 py-amqp:2.4.2
platform -> system:Linux arch:64bit
kernel version:4.18.0-1013-azure imp:CPython
loader -> celery.loaders.app.AppLoader
settings -> transport:amqp results:db+mysql+pymysql://celery:**@X.X.X.X:XXXX/celeryresults
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

* **Minimal Python Version**: N/A or Unknown
* **Minimal Celery Version**: N/A or Unknown
* **Minimal Kombu Version**: N/A or Unknown
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
pip freeze
amqp==2.4.2
asn1crypto==0.24.0
bcrypt==3.1.6
billiard==3.6.0.0
celery==4.3.0
cffi==1.12.2
Click==7.0
cryptography==2.6.1
Flask==1.0.2
Flask-Cors==3.0.7
itsdangerous==1.1.0
Jinja2==2.10
kombu==4.5.0
MarkupSafe==1.1.1
mysql-connector-python==8.0.15
netmiko==2.3.3
paramiko==2.4.2
protobuf==3.7.1
pyasn1==0.4.5
pycparser==2.19
PyMySQL==0.9.3
PyNaCl==1.3.0
pyserial==3.4
pytz==2018.9
PyYAML==5.1
scp==0.13.2
six==1.12.0
SQLAlchemy==1.2.14
textfsm==0.4.1
vine==1.3.0
Werkzeug==0.15.2
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
```

</p>
</details>

# Expected Behavior

With the following configured "result_extended" to true I expected to see more meta data in the backend results DB but I don't. I just see the standard results.

app = Celery('tasks',
             backend='db+mysql+pymysql://XXX:XXX@X.X.X.X:XXXX/celeryresults',
             broker='amqp://networkautomation:B2p2k6nb!@X.X.X.X/celery',
             result_extended=True)
# Actual Behavior

### PR DESCRIPTION
**Pull request by safwanrahman**
This fixes #5436 except `cassendra`. I do not have any experience with cassendra, so dont know how to fix that.

### CODE REVIEW

**Comment by thedrow** [comment]
What about backends such as [MongoDB](https://github.com/celery/celery/blob/master/celery/backends/mongodb.py#L181) which override `_store_result`?

**Comment by safwanrahman** [comment]
> What about backends such as [MongoDB](https://github.com/celery/celery/blob/master/celery/backends/mongodb.py#L181) which override `_store_result`?

I have changed the mongodb one also!

**Comment by thedrow** [comment]
It shouldn't be hard to add the same capability to the Cassandra backend.
You'll need to edit [these queries](https://github.com/celery/celery/blob/master/celery/backends/cassandra.py#L35).
Add a condition [here](https://github.com/celery/celery/blob/master/celery/backends/cassandra.py#L155), [here](https://github.com/celery/celery/blob/master/celery/backends/cassandra.py#L175) and [here](https://github.com/celery/celery/blob/master/celery/backends/cassandra.py#L161) to use the correct table creation, insert statement and read statement according to the configuration value.
CQL is very much like SQL in this regard.

**Comment by safwanrahman** [comment]
Thanks @geoffreybauduin. I am trying to run the test locally to fix the tests actually!

**Comment by sckhg1367** [comment]
hi guys,
do celery  Result Stores support MongoDB. Thanks you vvery much
image

**Comment by georgepsarakis** [comment]
@sckhg1367 yes, probably the documentation needs to be updated:

https://github.com/celery/celery/blob/master/celery/backends/mongodb.py

Would you like to open a PR perhaps to include all available result stores?

**Comment by sckhg1367** [comment]
Thanks @geoffreybauduin.

**Comment by akshjain83** [comment]
Any updates on this? This is a crucial fix and looks like we are pretty close here. Can someone please prioritize this?

**Comment by thedrow** [comment]
I'll work on it.

**Comment by safwanrahman** [comment]
Sorry I could not get time to work on this! @thedrow Can you let me know what do you think about failing tests?

**Comment by thedrow** [comment]
I'll rebase first and then let you know.

**Comment by akshjain83** [comment]
Great thanks guys for taking a look into this.

**Comment by thedrow** [comment]
I'm not sure I resolved the conflict correctly which is why I merged.
Can anyone take a look?

**Comment by auvipy** [comment]
seems only 1 failure

**Comment by safwanrahman** [comment]
@thedrow I think rebase is appropiate here. I will rebase it upon master and will take a look at the test failure.

**Comment by akshjain83** [comment]
Any updates here? Can someone please share the estimated timeline for releasing Milestone `4.4.0` ? Please let me know if I can help in some way, thanks.

**Comment by auvipy** [comment]
30th Nov

**Comment by safwanrahman** [comment]
@georgepsarakis @thedrow I have fixed the code to pass the tests. Can you take a look?

**Comment by auvipy** [comment]
green build and test coverage increase looks good!

**Comment by safwanrahman** [comment]
@auvipy Which tests you are referring?

**Comment by auvipy** [comment]
https://github.com/celery/celery/tree/master/t/unit/backends test_base, test_database, test_mongo

File: celery/backends/base.py [hunk 1]
```diff
@@ -351,6 +351,54 @@ def encode_result(self, result, state):
     def is_cached(self, task_id):
         return task_id in self._cache
 
+    def _get_result_meta(self, result,
```
Comment (auvipy) [inline celery/backends/base.py]: particularly this block and related changes tests

Comment (safwanrahman) [reply celery/backends/base.py]: @auvipy Do you mean writing new tests?

Comment (auvipy) [reply celery/backends/base.py]: yes some more test

**Comment by akshjain83** [comment]
@safwanrahman Are you working on the tests? I haven't contributed to this project before but I could try to help if you need any. Please let me know if we can still make it to the Milestone `4.4.0` release.

**Comment by auvipy** [comment]
> @safwanrahman Are you working on the tests? I haven't contributed to this project before but I could try to help if you need any. Please let me know if we can still make it to the Milestone `4.4.0` release.

you can push some tests on his forks branch :)

**Comment by akshjain83** [comment]
@auvipy @safwanrahman While I am touching this branch, just thought of checking with you guys if it would make sense to create a `__init__` method/constructor in this [class](https://github.com/celery/celery/blob/master/celery/backends/database/models.py#L50) as we have changes in this PR as:

https://github.com/celery/celery/pull/5661/files#diff-af101404f68c021a21d96325345ff894R121-R122

We can leverage this parent init via super: 
https://github.com/celery/celery/blob/master/celery/backends/database/models.py#L34-L35

**Comment by akshjain83** [comment]
> > @safwanrahman Are you working on the tests? I haven't contributed to this project before but I could try to help if you need any. Please let me know if we can still make it to the Milestone `4.4.0` release.
> 
> you can push some tests on his forks branch :)

@safwanrahman I few tests that I would like to push but I don't have permissions to push to your fork. Please let me know how you would like me this few changes.

**Comment by safwanrahman** [comment]
@akshjain83 Thanks for the commit, sorry I did not get time to write the tests myself. I have cherry-picked the commit and pushed to the branch.
@auvipy possible to merge?

**Comment by akshjain83** [comment]
No worries @safwanrahman Happy to contribute. @auvipy How does the timeline looks like for milestone `4.4.0` ? Any updated estimates?

**Comment by thedrow** [comment]
I'm hoping to make a release by the weekend.

**Comment by akshjain83** [comment]
Thanks @thedrow

**Comment by auvipy** [comment]
I released celery==4.4.0rc5

**Comment by akshjain83** [comment]
Thanks for the update @auvipy. Just curious, as per the comments [here](https://github.com/celery/celery/issues/5816#issuecomment-562859314), I am assuming that final release is the stable `4.4.0` release, right? And that would be released sometime before 2019?

**Comment by auvipy** [comment]
in a week yes

**Comment by akshjain83** [comment]
So I upgrade from `4.3.0` to `4.4.0rc5` version but it errors out. Looks like since there is no migration of tables in celery, if you have existing celery_taskmeta table with data, extended columns  won't be created.

**Comment by thedrow** [comment]
@akshjain83 We don't have migrations for that backend.
You'll have to add the columns yourself or contribute the migrations.

**Comment by akshjain83** [comment]
Sounds good, will start a separate issue for this.

**Pull request merged**